addons/mail/models/mail_channel.py

```diff
@@ -90,6 +105,113 @@ def default_get(self, fields):
              "Use this field anywhere a small image is required.")
     is_subscribed = fields.Boolean(
         'Is Subscribed', compute='_compute_is_subscribed')
+    moderation = fields.Boolean(string='Moderate this channel')
+    moderator_ids = fields.Many2many('res.users', 'mail_channel_moderator_rel', string='Moderators')
+    moderated_email_ids = fields.One2many('channel.moderated.emails', 'channel_id', string='Moderated Emails')
+    auto_notification = fields.Boolean(string="Automatic notification", help="People receive an automatic notification about their message being waiting for moderation.")
+    notification_message = fields.Text(string="Notification message")
+    send_guidelines = fields.Boolean(string="Send guidelines to new subscribers", help="Newcomers on this moderated channel will automatically receive the guidelines.", default=False)
+    guidelines = fields.Text(string="Guidelines")
+
+    @api.constrains('moderator_ids')
+    def _check_moderators_have_email_addresses(self):
+        if self.mapped('moderator_ids').filtered(lambda moderator: moderator.email == False):
+            raise ValidationError("Moderators must have an email address!")
+
+
+    @api.constrains('moderator_ids', 'channel_partner_ids', 'channel_last_seen_partner_ids')
+    def _check_moderators_are_channel_partners(self):
+        for channel in self:
+            if not (channel.mapped('moderator_ids.partner_id') <= channel.channel_partner_ids):
+                raise ValidationError("Moderators cannot unsubscribe from a channel they moderate!")
+
+    @api.constrains('moderation', 'email_send')
+    def _check_moderation_implies_email_send(self):
+        if self.filtered(lambda channel: channel.moderation and not channel.email_send):
+            raise ValidationError('Only email lists can be moderated!')
+
+    @api.onchange('moderator_ids')
+    def _add_moderators_to_channel_last_seen_partners(self):
+        ChannelPartner = self.env['mail.channel.partner']
+        for partner in (
+            self.mapped('moderator_ids.partner_id')
+            - self.mapped('channel_last_seen_partner_ids.partner_id')
```

Comment:
Isn't it channel_partner_ids ?

Reply:
Yes it is, but it doesn't work...
In the interface, I am changing channel_last_seen_partner_ids. So when I refer directly to channel_partner_ids the value of that has not yet been changed, only channel_last_seen_partner_ids. So that self.mapped('channel_last_seen_partner_ids.partner_id') gives what I what but not self.mapped('channel_partner_ids')!